teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,605 @@
+> also lead to confusion among users, which we should design for.
+
+Users need a way to safely and reliably substitute `Parameters` in `Scripts` in `Steps`
+of their `Tasks`. This would make it easier to use Tekton without learning advanced topics,
+while still using it reliably and safely.
+
+### Requirements
+
+<!--
+Describe constraints on the solution that must be met, such as:
+- which performance characteristics that must be met?
+- which specific edge cases that must be handled?
+- which user scenarios that will be affected and must be accommodated?
+-->
+
+1. **[Secure by default in v1][secure-by-default]**: Out of the box
+   *Tekton Pipelines* v1 should disallow common anti-patterns that
+   can render a default install insecure.
+1. **Easy to use**: Whatever replaces the direct use of `$(params.X)` in
+   `script` should be unsurprising and memorable to ease friction in user
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
+1. **Backwards-compatible in v1**. If a user already has `Tasks`
+   that use `$(params.X)` syntax in `script` then we should continue supporting
+   that usage in v1 resources. Ideally, Tekton Pipelines would warn the user
+   about it somehow.
+1. **(To be discussed) Provide an escape hatch in v1**. If a cluster
+   operator or admin decides that they want to continue allowing
+   `Parameter` variables to be directly injectable to `script` fields of
+   `Steps` there should be some way for them to configure that.
+
+## Proposal - Implicitly Project `Parameters` as Environment Variables in `Steps` (Language-Agnostic)
+We have projected `Parameters` as environment variables in the `git-clone` Task in the
+Catalog and in our own plumbing repository. We can use this approach, and simplify the
+syntax by implicitly projecting `Parameters` as environment variables in `Steps`.
+The environment variable should be prefixed with `PARAM_` to make it clear and predictable.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone ${$(params.url)} # <--- needs to be manually updated w/ additional ${*} wrapping for shell usage
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone ${PARAM_URL} 
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_URL
+      value: $(params.url)
+    script: |
+      git clone ${PARAM_URL}
+```
+### Discussion
+This approach mutates the specification, but it is something we're already doing
+with implicit `Parameters` per [TEP-0023][tep-0023].  This approach removes the insecure
+default behaviour but in doing so unfortunately loses backwards-compatibility requiring users
```

Comment:
Thanks @aaron-prindle for putting putting up the discussion. I noticed there are already discussions at https://github.com/tektoncd/community/pull/1077#discussion_r1344697979 on using feature gate to turn off the default behaviour. But I am afraid the existing feature has now been regarded as stable, and we might have to comply with the [API compatibility policy](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md) for its deprecation cycle to keep both options available to users.

Reply:
Also noticed in 
- "During the inital creation of this feature, the user experience will not change as the original UX/functionality will be preserved." 
- "At a later time or in an updated `apiVersion` (v2*),"

These all SGTM. Wondering if we would like to call out explicitly that we'd wait till the end of the 9-month deprecation period?

Reply:
> Wondering if we would like to call out explicitly that we'd wait till the end of the 9-month deprecation period?
@JeromeJu where would this one apply ?

Reply:
I've updated the text here to explicitly reference the 9 mo deprecation policy for any v1.X changes

Reply:
As noted here and in some other comments, the doc was initially unclear around the backwards compatibility guaratees.  The propsed solution will be backwards compatible for the 9 mo deprectation window and will have an opt-in feature flag for users to modify this if they desire (assuming folks here agree to that during the implementation).